### PR TITLE
ARROW-6398: [C++] Consolidate ScanOptions and ScanContext

### PR DESCRIPTION
**Pull request by bkietz**
Currently ScanOptions has two distinct responsibilities: it contains the data selector (and eventually projection schema) for the current scan and it serves as the base class for format specific scan options. This makes providing scan options for more than a single format impossible (as the resulting merged options would for example need to inherit both JsonScanOptions and ParquetScanOptions).

This patch removes ScanOptions and promotes FileScanOptions to the abstract base class for format specific scan options.

ScanContext is now the root container of scan state: contains the data selector, projection schema, and a vector of FileScanOptions

### CODE REVIEW

File: cpp/src/arrow/dataset/scanner.h [hunk 1]
```diff
@@ -29,26 +29,61 @@
...
-// TODO(wesm): API for handling of post-materialization filters. For
-// example, if the user requests [$col1 > 0, $col2 > 0] and $col1 is a
-// partition key, but $col2 is not, then the filter "$col2 > 0" must
-// be evaluated in-memory against the RecordBatch objects resulting
-// from the Scan
+  virtual ~FileScanOptions() = default;
+};
 
-class ARROW_DS_EXPORT ScanOptions {
+struct ARROW_DS_EXPORT ScanContext final {
```
Comment (pitrou) [inline cpp/src/arrow/dataset/scanner.h]: I think this should still be named `ScanOptions`. Perhaps you can use `ScanContext` for some opaque internal state (if necessary).

File: cpp/src/arrow/dataset/scanner.h [hunk 2]
```diff
@@ -29,26 +29,61 @@
...
+    pool_ = p;
+    return *this;
+  }
+
+  /// format specific options
+  const std::vector<std::shared_ptr<FileScanOptions>>& options() const {
+    return options_;
+  }
+
+  ScanContext& AddFileScanOptions(std::shared_ptr<FileScanOptions> opts) {
```
Comment (pitrou) [inline cpp/src/arrow/dataset/scanner.h]: `file_options`?

File: cpp/src/arrow/dataset/scanner.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/scanner.h]: Why should it be named ScanOptions? It isn't optional, unlike the potentially absent FileScanOptions. Furthermore that name implies that FileScanOptions inherits ScanOptions, which is no longer the case. If ScanContext isn't a good name then what about ScanState?

File: cpp/src/arrow/dataset/scanner.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/scanner.h]: Alright

File: cpp/src/arrow/dataset/scanner.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/scanner.h]: Because those are options, simply.

Comment (bkietz) [reply cpp/src/arrow/dataset/scanner.h]: But we also have `compute::FunctionContext` which is required by the kernels, and may optionally be configured with a different memory pool

Comment (pitrou) [reply cpp/src/arrow/dataset/scanner.h]: Lots of things can be configured with a memory pool. But `FunctionContext` doesn't hold any other options; it's a container for kernel state execution, not user-provided preferences.

Comment (fsaintjacques) [reply cpp/src/arrow/dataset/scanner.h]: Both argument are valid, it holds a MemoryPool like FunctionContext, but it has options like other kernel. I'm biased toward ScanOptions as it holds important options like the projected schema and the filters.

Comment (bkietz) [reply cpp/src/arrow/dataset/scanner.h]: Alright, I'll rename back to ScanOptions.

File: cpp/src/arrow/dataset/scanner.h [hunk 3]
```diff
@@ -29,31 +29,73 @@
...
+/// Container for scan state
+struct ARROW_DS_EXPORT ScanOptions final {
  public:
-  virtual ~ScanOptions() = default;
+  ScanOptions() = default;
 
+  /// Filters
   const std::shared_ptr<DataSelector>& selector() const { return selector_; }
 
+  ScanOptions& selector(std::shared_ptr<DataSelector> s) {
```
Comment (pitrou) [inline cpp/src/arrow/dataset/scanner.h]: Not sure about the naming. Should this kind of method be called `set_selector`? @wesm

**Comment by pitrou** [comment]
Can you rebase in order to fix the GLib / C issues?

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
While working on discovery, I noted a problem with this change. Let's distinguish between 2 phases.

- **Discovery**: in the discovery phase, directories/services will be crawled to find available files/fragments. Schema may or not yet be fixed, but the discovery phase allows inferring/enforcing the DataSource common schema **before** materializing the DataSource. In order to do this, it needs to read/peek files, and thus it needs CsvReaderOptions, ParquetReaderOptions, etc... The discovery phase needs to create a DataSource with said "read" options + schema _fixed_ in the DataSource.

- **Scan**: in the scan phase, each DataSource is iterated over to concatenate their fragments. The caller controls the filter clause (which fragments to skip), and the projected (subset or superset) schema.

The discovery options (how to open a file, connect url, reconciliation schema) should not be passed by the caller at Scan phase, they're fixed in the DataSource. The (Scan) user shouldn't care about CsvReadOption (that's the goal of the dataset project, hide those details).

I think the existing classes (ScanOptions, ScanContext) convey this but with unclear naming, e.g. it would probably make more sense to rename and re-organize as follow:

- `ScanContext` -> `ScanOptions(proj: Schema, selector: Filter, ctx: (MemoryPool...))`
- `ScanOptions` -> `DataSourceOptions(reconcile: Schema, format: FileFormat, options: FileFormatOptions(CsvReader...))`.

The DataFragment constructor would take DataSourceOptions, while FileFragment::Scan would take the ScanOptions as parameter.

**Comment by bkietz** [comment]
Closing for now, will try again later

**Pull request closed without merging**